cli/src/cmd_disk.rs

```diff
@@ -4,27 +4,457 @@
+use oxide_api::types::DiskSource;
+use oxide_api::types::FinalizeDisk;
+use oxide_api::types::ImageCreate;
+use oxide_api::types::ImageSource;
+use oxide_api::types::ImportBlocksBulkWrite;
+use oxide_api::types::Name;
+use oxide_api::types::NameOrId;
+use oxide_api::types::Snapshot;
+use oxide_api::Client;
+use oxide_api::ClientDisksExt;
+use oxide_api::ClientImagesExt;
+use oxide_api::ClientSnapshotsExt;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use tokio::sync::mpsc;
 
+/// Import a file to a disk
+///
+/// Upload the contents of a file to a disk, optionally creating a snapshot
+/// and/or an image.
 #[derive(Parser, Debug, Clone)]
 #[clap(verbatim_doc_comment)]
 #[clap(name = "import")]
-pub struct CmdDiskImport {}
+pub struct CmdDiskImport {
+    /// Name or ID of the project
+    #[clap(long)]
+    project: String,
+
+    #[clap(long)]
+    description: String,
+
+    /// Path to the file to import
+    #[clap(long)]
+    path: String,
+
+    /// How many upload threads to run. Disk size must be divisible by this
+    /// number.
+    #[clap(long, default_value = "1")]
+    threads: usize,
+
+    /// The name of the disk to create
+    #[clap(long)]
+    disk_name: Name,
+
+    /// The size of the disk to create. If unspecified, the size of the file
+    /// will be used, rounded up to the nearest GB.
+    #[clap(long)]
+    disk_size: Option<u64>,
+
+    /// The desired size of the disk's blocks.
+    #[clap(long)]
+    disk_block_size: Option<i64>,
+
+    /// If supplied, create a snapshot with this name.
+    #[clap(long)]
+    snapshot_name: Option<Name>,
+
+    /// If supplied, create an image from the snapshot of this disk. Requires
+    /// snapshot name to be supplied.
+    #[clap(long)]
+    image_name: Option<Name>,
+
+    /// The description for the image created out of the snapshot of this disk.
+    #[clap(long)]
+    image_description: Option<String>,
+
+    /// The OS of this image (eg. Debian)
+    #[clap(long)]
+    image_os: Option<String>,
+
+    /// The version of this image (eg. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
+    #[clap(long)]
+    image_version: Option<String>,
+}
+
+/// Return a correct disk size for the path and size arguments
+fn get_disk_size(path: &str, size: Option<u64>) -> Result<u64> {
+    const ONE_GB: u64 = 1024 * 1024 * 1024;
+
+    let disk_size = if let Some(size) = size {
+        size
+    } else {
+        std::fs::metadata(path)?.len()
+    };
+
+    // Nexus' disk size minimum is 1 GB
+    let disk_size = std::cmp::max(ONE_GB, disk_size);
+
+    // Nexus only supports disks whose size is a multiple of 1 GB
+    let disk_size = if disk_size % ONE_GB != 0 {
+        let rounded_down_gb: u64 = disk_size - disk_size % ONE_GB;
+        assert_eq!(rounded_down_gb % ONE_GB, 0);
+        rounded_down_gb + ONE_GB
+    } else {
+        disk_size
+    };
+
+    Ok(disk_size)
+}
+
+// Upload to Nexus in 512k byte chunks
+const CHUNK_SIZE: u64 = 512 * 1024;
```

Comment:
This seems really small!

Reply:
Ah. It comes from Crucible: https://github.com/oxidecomputer/crucible/blob/0b34a53b57038a7804ac69ab90400539bd1961e8/pantry/src/pantry.rs#L33